File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeHttpResponseParser.java
package com.braintreepayments.api;

import androidx.annotation.VisibleForTesting;

import java.net.HttpURLConnection;

/**
 * Class that handles parsing http responses for {@link BraintreeHttpClient}.
 */
class BraintreeHttpResponseParser implements HttpResponseParser {

    private final HttpResponseParser baseParser;

    BraintreeHttpResponseParser() {
        this(new BaseHttpResponseParser());
    }

    @VisibleForTesting
    BraintreeHttpResponseParser(HttpResponseParser baseParser) {
        this.baseParser = baseParser;
    }

    /**
     * @param responseCode the response code returned when the http request was made.
     * @param connection the connection through which the http request was made.
     * @return the body of the http response.
     */
    @Override
    public String parse(int responseCode, HttpURLConnection connection) throws Exception {
        try {
            return baseParser.parse(responseCode, connection);
        } catch (AuthorizationException | UnprocessableEntityException e) {
            if (e instanceof AuthorizationException) {
                String errorMessage = new ErrorWithResponse(403, e.getMessage()).getMessage();
                throw new AuthorizationException(errorMessage);
            } else {
                throw new ErrorWithResponse(422, e.getMessage());
            }
        }
    }
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/BraintreeHttpResponseParser.kt
package com.braintreepayments.api

import java.net.HttpURLConnection

/**
 * Class that handles parsing http responses for [BraintreeHttpClient].
 */
internal class BraintreeHttpResponseParser(
    private val baseParser: HttpResponseParser = BaseHttpResponseParser()
) : HttpResponseParser {

    /**
     * @param responseCode the response code returned when the http request was made.
     * @param connection the connection through which the http request was made.
     * @return the body of the http response.
     */
    @Throws(Exception::class)
    @Suppress("SwallowedException")
    override fun parse(responseCode: Int, connection: HttpURLConnection): String = try {
        baseParser.parse(responseCode, connection)
    } catch (e: AuthorizationException) {
        val errorMessage = ErrorWithResponse(AUTH_ERROR_CODE, e.message).message
        throw AuthorizationException(errorMessage)
    } catch (e: UnprocessableEntityException) {
        throw ErrorWithResponse(UNPROCESSABLE_ENTITY_ERROR_CODE, e.message)
    }

    companion object {
        private const val AUTH_ERROR_CODE = 403
        private const val UNPROCESSABLE_ENTITY_ERROR_CODE = 422
    }
}
